File: frontend/src/services/heatmap.ts

import { apiClient } from './api';

export interface TimeSlot {
  weekday: string;
  hour: number;
  avg_pnl: number;
  total_pnl: number;
  trade_count: number;
  win_rate: number;
}

export interface SymbolStats {
  symbol: string;
  total_trades: number;
  total_pnl: number;
  avg_pnl: number;
  win_rate: number;
  profit_factor: number;
  best_trade: number;
  worst_trade: number;
  total_volume: number;
  avg_volume: number;
  trading_days: number;
  avg_trades_per_day: number;
  direction_bias: string;
  long_trades: number;
  short_trades: number;
  consistency_score: number;
}

export interface HeatmapInsight {
  type: string;
  message: string;
  data: any;
}

export interface HeatmapRecommendation {
  type: string;
  priority: 'high' | 'medium' | 'low';
  title: string;
  description: string;
  expected_impact: string;
}

export interface TimeHeatmapData {
  total_pnl_matrix: Record<string, number[]>;
  avg_pnl_matrix: Record<string, number[]>;
  trade_count_matrix: Record<string, number[]>;
  win_rate_matrix: Record<string, number[]>;
  best_time_slots: TimeSlot[];
  worst_time_slots: TimeSlot[];
  total_trading_hours: number;
}

export interface SymbolStatsData {
  symbols: SymbolStats[];
  summary: {
    total_symbols: number;
    profitable_symbols: number;
    unprofitable_symbols: number;
    best_symbol: string | null;
    worst_symbol: string | null;
    most_traded_symbol: string | null;
    avg_symbols_per_day: number;
  };
}

export interface HeatmapData {
  time_heatmap: TimeHeatmapData;
  symbol_stats: SymbolStatsData;
  insights: {
    time_insights: HeatmapInsight[];
    symbol_insights: HeatmapInsight[];
    recommendations: HeatmapRecommendation[];
  };
  metadata: {
    total_trades: number;
    date_range: {
      start: string | null;
      end: string | null;
    };
    symbols_analyzed: number;
  };
}

export interface HeatmapResponse {
  success: boolean;
  data: HeatmapData;
}

export const heatmapService = {
  async getPerformanceHeatmap(
    startDate?: string,
    endDate?: string
  ): Promise<HeatmapData> {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);
    
    const response = await apiClient.get<HeatmapResponse>(
      `/analytics/heatmap?${params.toString()}`
    );
    
    if (!response.data.success) {
      throw new Error('Failed to fetch heatmap data');
    }
    
    return response.data.data;
  },

  async getTimeHeatmapOnly(
    startDate?: string,
    endDate?: string
  ): Promise<{
    time_heatmap: TimeHeatmapData;
    insights: HeatmapInsight[];
    metadata: HeatmapData['metadata'];
  }> {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);
    
    const response = await apiClient.get(
      `/analytics/heatmap/time?${params.toString()}`
    );
    
    if (!response.data.success) {
      throw new Error('Failed to fetch time heatmap data');
    }
    
    return response.data.data;
  },

  async getSymbolPerformanceOnly(
    startDate?: string,
    endDate?: string
  ): Promise<{
    symbol_stats: SymbolStatsData;
    insights: HeatmapInsight[];
    metadata: HeatmapData['metadata'];
  }> {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);
    
    const response = await apiClient.get(
      `/analytics/heatmap/symbols?${params.toString()}`
    );
    
    if (!response.data.success) {
      throw new Error('Failed to fetch symbol performance data');
    }
    
    return response.data.data;
  }
};

File: src/backend/api/v1/analytics/heatmap.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, Dict, Any
from datetime import datetime

from backend.core.db.session import get_db
from backend.api.deps import get_current_user
from backend.services.analytics.heatmap import HeatmapAnalyticsService

router = APIRouter(tags=["heatmap-analytics"])

def get_heatmap_service(db: Session = Depends(get_db)) -> HeatmapAnalyticsService:
    return HeatmapAnalyticsService(db)

@router.get("/heatmap")
async def get_performance_heatmap(
    start_date: Optional[datetime] = Query(None, description="Start date filter"),
    end_date: Optional[datetime] = Query(None, description="End date filter"),
    current_user: dict = Depends(get_current_user),
    service: HeatmapAnalyticsService = Depends(get_heatmap_service)
) -> Dict[str, Any]:
    """
    Get comprehensive performance heatmap data including:
    - Time-based heatmap (hour × weekday)
    - Symbol performance statistics  
    - Actionable insights and recommendations
    """
    try:
        heatmap_data = await service.generate_heatmap_data(
            user_id=current_user["user_id"],
            start_date=start_date,
            end_date=end_date
        )
        
        return {
            "success": True,
            "data": heatmap_data
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500, 
            detail=f"Heatmap analysis failed: {str(e)}"
        )

@router.get("/heatmap/time")
async def get_time_heatmap_only(
    start_date: Optional[datetime] = Query(None, description="Start date filter"),
    end_date: Optional[datetime] = Query(None, description="End date filter"),
    current_user: dict = Depends(get_current_user),
    service: HeatmapAnalyticsService = Depends(get_heatmap_service)
) -> Dict[str, Any]:
    """Get only time-based heatmap data for focused analysis"""
    try:
        heatmap_data = await service.generate_heatmap_data(
            user_id=current_user["user_id"],
            start_date=start_date,
            end_date=end_date
        )
        
        return {
            "success": True,
            "data": {
                "time_heatmap": heatmap_data["time_heatmap"],
                "insights": heatmap_data["insights"]["time_insights"],
                "metadata": heatmap_data["metadata"]
            }
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500, 
            detail=f"Time heatmap analysis failed: {str(e)}"
        )

@router.get("/heatmap/symbols")
async def get_symbol_performance_only(
    start_date: Optional[datetime] = Query(None, description="Start date filter"),
    end_date: Optional[datetime] = Query(None, description="End date filter"),
    current_user: dict = Depends(get_current_user),
    service: HeatmapAnalyticsService = Depends(get_heatmap_service)
) -> Dict[str, Any]:
    """Get only symbol performance data for focused analysis"""
    try:
        heatmap_data = await service.generate_heatmap_data(
            user_id=current_user["user_id"],
            start_date=start_date,
            end_date=end_date
        )
        
        return {
            "success": True,
            "data": {
                "symbol_stats": heatmap_data["symbol_stats"],
                "insights": heatmap_data["insights"]["symbol_insights"],
                "metadata": heatmap_data["metadata"]
            }
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500, 
            detail=f"Symbol performance analysis failed: {str(e)}"
        )
